python/tskit/stats.py

```diff
@@ -163,3 +167,506 @@ def r2_matrix(self):
+        # nodes be updated for trio first coalescences?
+
+        # nodes where descendant subtree has been altered
+        modified = {i.parent for i in edge_diff.edges_in}
+        for i in copy.deepcopy(modified):
+            while not tree.parent(i) is tskit.NULL and not tree.parent(i) in modified:
+                i = tree.parent(i)
+                modified.add(i)
+
+        # recalculate weights for current tree
+        for i in removed:
+            running_weights[i, :] = 0
+        for i in modified:
+            running_weights[i, :] = self.weight_func(i, tree, self.sample_sets)
+        self.weight_func_evals += len(modified)
+
+    def _build_ecdf_table_for_window(
+        self, left, right, tree, edge_diffs, running_weights
+    ):
+        """
+        Construct ECDF table for genomic interval [left, right]. Update ``tree``,
+        ``edge_diffs``, and ``running_weights`` for input for next window. Trees are
+        counted as belonging to any interval with which they overlap, and thus
+        can be used in several intervals. Thus, the concatenation of ECDF
+        tables across multiple intervals is not the same as the ECDF table
+        for the union of those intervals. Trees within intervals are chunked
+        into roughly equal-sized blocks for bootstrapping.
+        """
+
+        assert tree.interval.left <= left and right > left
+
+        # assign trees in window to equal-sized blocks with unique id
+        _tree = tree.copy()
+        if right >= _tree.tree_sequence.sequence_length:
+            _tree.last()
+        else:
+            # _tree.seek(right) won't work if `right` is recomb breakpoint
+            while _tree.interval.right < right:
+                _tree.next()
+        tree_idx = np.arange(tree.index, _tree.index + 1) - tree.index
+        tree_offset = tree.index
+        num_blocks = min(self.num_blocks, len(tree_idx))
+        tree_blocks = np.floor_divide(num_blocks * tree_idx, len(tree_idx))
+
+        # calculate span weights
+        # TODO: if bootstrapping, does block span need to be tracked
+        # and used to renormalise each replicate?
+        _tree = tree.copy()
+        tree_span = [min(_tree.interval.right, right) - max(_tree.interval.left, left)]
+        while _tree.index < tree_offset + tree_idx[-1]:
+            _tree.next()
+            tree_span.append(
+                min(_tree.interval.right, right) - max(_tree.interval.left, left)
+            )
+        tree_span = np.array(tree_span) / sum(tree_span)
+
+        # storage if using single window, block for entire tree sequence
+        buffer_size = self.buffer_size
+        table_size = buffer_size
+        time = np.zeros(table_size)
+        block = np.zeros(table_size, "int32")
+        weights = np.zeros((table_size, self.num_weights))
+
+        # assemble table of coalescence times in window
+        indices = np.zeros(tree.tree_sequence.num_nodes, "int32") - 1
```

Comment:
We usually use ``dtype=np.int32`` - saves a little bit of string looking up, presumably. Bit more explicit to use the named param for dtype also.